### PR TITLE
[WIP] Autogenerate state selectors entry file

### PR DESCRIPTION
**Pull request by retrofox**
This PR attempts to suggest a way auto-generate the `index.js` of the state-selectors.
This index.js file is just an entry point for all files located into the `./client/state/selectors` folder.

**IMPORTANT: As far as I am concerned I consider this patch as an experiment with an uncertain future.**

### How does it work?
The generation process is started after the calypso starts up. So the idea is when either a new file is added into the selectors folder or when an existent file is removed from there, the process will re-generate the `index.js` file updating the exported modules.
The `generator` is watching the selectors folder only under these conditions: add/remove a file.

For instance, [this commit](https://github.com/Automattic/wp-calypso/pull/10804/commits/d0d8a0958f904349f448d3a5fdd7bb476eb32810) has been autogenerated.

### Following the generation process

The generation process can be followed taking a look in the terminal. As soon as Calypso starts the log line is shown there.

Below is an animation which shows how the `isRequestingTimezones` selector is added.

state-selectors-generator-08

### Generate

Although the generation process will be working on the background you would generate process through of a npm script:

```cli
> npm run generate-state-selectors
```

### Workflow

The workflow to add a new selector is simple. Create the new file into the selectors folder paying special attention to the name of the exported module and the name of the file. It must be consistent regarding the kebab - camelCase conversion.
This isn't something new since this convention has been established before to this patch. The new thing here is now you haven't to be concerned on declaring this selector into the index.js file.

For instance, if you create a `is-requesting-timezones.js` selector file, the exported module should be `isRequestingTimezones`.

Although the index.js file is generated automatically it doesn't mean that we won't have to commit its changes. Commit the modifications as usual.

### Testing

After pull the patch should ensure clean your local installation and enable the `state-selectors-generator` feature flag.

```
> make clean
> make distclean
> make install
> ENABLE_FEATURES=state-selectors-generator make run
```

Add/remove files to/from `./client/state/selectors/` folder paying attention how the index.js right there is updated in the meanwhile.

### CODE REVIEW

File: server/state-selectors-generator/index.js [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+				const exportLine = `export ${ moduleName } from '${ importBasePath }/${ file }';`;
+				indexContent += exportLine + '\n';
+
+				if ( counter === total ) {
+					saveIndexFile( indexContent );
+				}
+			}
+		} );
+	} );
+}
```
Comment (dmsnell) [inline server/state-selectors-generator/index.js]: just as a note, I think we could benefit by further breaking apart these functions into separate parts of this file. you've done a nice job modularizing the code! let's take that one step further and create a separate function for handling each individual file

File: server/state-selectors-generator/index.js [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+		let counter = 0;
+
+		files.forEach( file => {
+			counter++;
+			if ( ! file.match( blacklist ) ) {
+				file = file.replace( /\.jsx?/g, '' );
+
+				const moduleName = file.replace( /-([a-z])/g, function( str ) {
+					return str[ 1 ].toUpperCase();
+				} );
```
Comment (dmsnell) [inline server/state-selectors-generator/index.js]: what's the purpose of these lines? I can't follow the logic inside the RegExp

File: server/state-selectors-generator/index.js [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+	let indexContent = '';
+
+	fs.readdir( selectorsGlobalPath, ( err, files ) => {
+		const total = files.length - 1;
+		let counter = 0;
+
+		files.forEach( file => {
+			counter++;
+			if ( ! file.match( blacklist ) ) {
+				file = file.replace( /\.jsx?/g, '' );
```
Comment (dmsnell) [inline server/state-selectors-generator/index.js]: `g`? is there a reason for setting this flag?

**Comment by dmsnell** [comment]
sorry @retrofox I left a partial review. thanks for putting in this work! I'll try and get back to it later.

File: server/state-selectors-generator/script-generate.js [hunk 4]
```diff
@@ -0,0 +1,2 @@
+const stateSelectorsGenerator = require( './' );
+stateSelectorsGenerator.generateIndexFile();
```
Comment (dmsnell) [inline server/state-selectors-generator/script-generate.js]: do we not need an empty line here?

File: server/state-selectors-generator/index.js [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+};
+
+/*
+ * Start the generator process
+ */
+exports.watch = function() {
+	log( 'start to whatch ...' );
+	watcher
+		.on( 'add', addFileSelector )
+		.on( 'unlink', removeFileSelector );
```
Comment (dmsnell) [inline server/state-selectors-generator/index.js]: nice job listening for deletions too.

does this handle renames?

Comment (retrofox) [inline server/state-selectors-generator/index.js]: I guess so. Let me test it. In theory, it should be supported as well as symlinking.

Comment (retrofox) [inline server/state-selectors-generator/index.js]: It works for rename too. :-D

File: server/state-selectors-generator/script-generate.js (hunk 4 above)
Comment (retrofox) [inline server/state-selectors-generator/script-generate.js]: It seems to be already there. 👻

Comment (retrofox) [inline server/state-selectors-generator/script-generate.js]: Actually, it isn't.

**Comment by dmsnell** [comment]
I'm trying to figure out how much value we have in testing this. it seems like things would inherently fail if this fails, but maybe it wouldn't be that hard to rearrange some of the functions so that they are more easily testable. for example, we could create a string or array with mock directory contents and feed it into the function which creates the import file. this would let us test the pattern matching and file filtering and formats without doing any fancy mocking work.

**Comment by retrofox** [comment]
> I'm trying to figure out how much value we have in testing this.

Definitely, we have to test this feature as much as possible.

> it seems like things would inherently fail if this fails...

No more than now. In the worst of cases, the application won't be able to require a module, etc. the same situation occurs if you aren't consistent between export/importing the module.

Some considerations:

* it only works in `development` environment.
* Although the index.js file is auto generated we will have to commit the changes. We aren't automatizing the most important task here.

Add tests is imperative and matter of time IMO. I'd just to know if it's a valid approach.

**Comment by dmsnell** [comment]
> Although the index.js file is auto generated we will have to commit the changes. We aren't automatizing the most important task here.

this is something I forgot to bring up too. I'm not a fan of committing generated code and I wondered what we might be able to do to perform this without checking it in.

by the way, great work here @retrofox

**Comment by retrofox** [comment]
> > Although the index.js file is auto generated we will have to commit the changes. We aren't automatizing the most important task here.

> this is something I forgot to bring up too. I'm not a fan of committing generated code and I wondered what we might be able to do to perform this without checking it in.

We can, but I suggest advance step by step. Auto Generate the code in prod is something that, at the very least, scares me. It's completely related to add tests to avoid this situation of course.

**Comment by dmsnell** [comment]
> Auto Generate the code in prod

the way I thought @aduth explained it however is that this file isn't even used in production, that the babel transform means we bypass it

**Comment by retrofox** [comment]
I'll close this pair of PR. We can take a look later.

**Pull request closed without merging**